utils.py

```diff
@@ -0,0 +1,190 @@
+        "asset",
+        "capex_variable",
+        "opex_variable",
+        "lifetime",
+        "co2_emissions",
+        "density",
+        "energy_density",
+        "volumetric_cost",
+    ):
+        if col_name not in headers:
+            raise ValueError(
+                f"The column header '{col_name}' is missing in your input file {filename} under the '{name}' sheet"
+            )
+
+    name = "timeseries"
+    headers = [c.value for c in wb[name][1]]
+
+    for hd in ("CriticalDemand", "Demand", "SolarGen"):
+        if hd not in headers:
+            raise ValueError(
+                f"The column header '{hd}' is missing in your input file {filename} under the '{name}' sheet"
+            )
+
+    df_timeseries = pd.DataFrame(tuple(wb[name].values)[1:], columns=headers)
+
+    name = "settings"
+    headers = [c.value for c in wb[name][1]]
+
+    for hd in ("param", "value"):
+        if hd not in headers:
+            raise ValueError(
+                f"The column header '{hd}' is missing in your input file {filename} under the '{name}' sheet"
+            )
+    df = pd.DataFrame(tuple(wb[name].values)[1:], columns=headers)
+    df = df[df.columns.dropna()].set_index("param")
+    df_settings = df["value"]
+
+    # compute the epc needed for oemof investments if not provided
+    if "annuity" not in df_costs.columns:
+        wacc = df_settings.wacc
+        project_lifetime = df_costs.lifetime.project
+        # correspond to equation (6)
+        df_costs["annuity"] = df_costs.apply(
+            lambda x: annualized_capex(
+                x.capex_variable, project_lifetime, x.lifetime, wacc
+            )
+            + x.opex_fix,
+            axis=1,
+        )
+
+    name = "sensitivity"
+    headers = [c.value for c in wb[name][1]]
+
+    for hd in ("category", "variable_name", "min_val", "max_val", "step"):
+        if hd not in headers:
+            raise ValueError(
+                f"The column header '{hd}' is missing in your input file {filename} under the '{name}' sheet"
+            )
+    df_sensitivity = pd.DataFrame(tuple(wb[name].values)[2:], columns=headers).dropna()
+    if df_sensitivity.empty is False:
+        return df_costs, df_timeseries, df_settings, df_sensitivity
+    else:
+        return (
+            df_costs,
+            df_timeseries,
+            df_settings,
+        )
+
+
+def encode_image_file(img_path):
+    """Encode image files to load them in the dash layout under img html tag
+
+    Parameters
+    ----------
+    img_path: str
+        path to the image file
+
+    Returns
+    -------
+    encoded_img: bytes
+        encoded bytes of the image file
+
+    """
+
+    try:
+        with open(img_path, "rb") as ifs:
+            encoded_img = base64.b64encode(ifs.read())
+    except FileNotFoundError:
+        encoded_img = base64.b64encode(bytes())
+    return encoded_img
+
+
+def annualized_capex(
+    investment_t0, project_lifetime, asset_lifetime=None, wacc=0.05, tax=0
```

Comment:
To which  `investment_t0` refers to?

Reply:
It looks like it is added to the `capex_from_investment` but not defined.

Reply:
you can also look where capex_from_investment is used in the code with ctrl+f